### ISSUE TITLE
Missing expansions

### ISSUE THREAD

**Issue by Kaligula0** [tadeaustria/EverdellScore#30]
* [Everdell: Legends](https://boardgamegeek.com/boardgameexpansion/297972/everdell-legends)
* [Everdell: Extra! Extra!](https://boardgamegeek.com/boardgameexpansion/297973/everdell-extra-extra)

**Comment by tadeaustria** [comment]
I am aware of these mini expansion, but I am a bit ambivalent about them. Some of the Legends cards, are also included in Mistwood expansion.
But yes, I agree, to be complete also the [Kindergarden](https://boardgamegeek.com/boardgameexpansion/382190/everdell-kindergarten-promo-cards) and the [Rugwort Pack](https://boardgamegeek.com/boardgameexpansion/261142/everdell-rugwort-pack) should be also included.

**Comment by Kaligula0** [comment]
Well, we always play with these mini expansions, so we cannot use your tool 🤷‍♂️ :(

I looked at the [BoardGameGeek Mistwood images](https://boardgamegeek.com/image/7160626/everdell-mistwood) and I see that all its 10 Legends cards are completely different from the Legends mini expansion.

I've heard about Rugwort Pack, but that's 3 cards for solo mode, which is not (yet) available in your tool, AFAIK. It can wait for later.

Kindergarten is just promo pack, so it's not widely available – look at the [number of owners](https://boardgamegeek.com/boardgame/199792/everdell/expansions?pageid=2&sort=numowned) (377 on BGG for now) versus expansions (6k-17k) and mini expansions (3.7k). Ideally it could be included, but well, can wait to be the last.

**Comment by tadeaustria** [comment]
Yes, I see. And I was wondering what I should do with these mini expansions, because I know they exist, but I don't own any of them. They are not available in my country so far.

I do not care about the priority, my biggest concern is that I implement something wrong, when iself don't use it.

I will create a suggestion for these in the near future.

**Comment by Kaligula0** [comment]
*Extra! Extra!* is just 15 new regular cards (of 6 types). You can see them in [BGG gallery](https://boardgamegeek.com/boardgameexpansion/297973/everdell-extra-extra/images).

*Legends* are also just 10 additional cards. You can see them in the [BGG gallery](https://boardgamegeek.com/boardgameexpansion/297972/everdell-legends/images). They're named *Legendary cards* because they are more powerful versions of a few normal cards (e.g. *Fynn Nobletail* can replace *King*), but in general they work as normal. The diferences are:
- they do not count to the city cards quota
- each player has a chance to play only 1 *Legendary Critter* and 1 *Legendary Construction* to his city

**Comment by tadeaustria** [comment]
Yes, I am aware of Legendary card rules, as they are already implemented in Mistwood. Nevertheless it requires special implementation for some, so its not _"just"_ cards...
Btw Legendary cards indeed do count to the city cards quota, but they allow you to have an additional space, which is not quite the same, once your town is full.

**Comment by Kaligula0** [comment]
> Btw Legendary cards indeed do count to the city cards quota, but they allow you to have an additional space, which is not quite the same, once your town is full.

No, see https://boardgamegeek.com/thread/2727273/article/38520138#38520138

**Comment by tadeaustria** [comment]
Reading that, I agree. But anyhow, Iabsolutely do not get, why legendary have all of this extra rule of _adding an extra space to town_. They could've just said, legendary do not count to town limit and done, as every player just has two legendary cards from the beginning. Would've made my implementation much easier...

**Comment by Kaligula0** [comment]
I learned that there are 2 cards that count *spaces you have filled in your city* – *Ever Wall* and *Mayor*. Whatever they are (Critters? Constructions? Events?) and from whichever expansions they are… There's a guy that knows more – @Dajakk.

**Comment by Dajakk** [comment]
> They are not available in my country so far.

They are part of [Glimmergold Pack ](https://boardgamegeek.com/boardgameexpansion/289405/everdell-glimmergold-pack)(2019), which is supposedly available in German

**Comment by Dajakk** [comment]
@Kaligula0 
I've added PR for extra extra. I've tested it some, but if you have some time it would be nice to get someone else take a look at that. DM me for details how to run it on localhost

**Comment by tadeaustria** [comment]
> > They are not available in my country so far.
> 
> They are part of [Glimmergold Pack ](https://boardgamegeek.com/boardgameexpansion/289405/everdell-glimmergold-pack)(2019), which is supposedly available in German

Yes, I know that. But I wasn't able to get my hands on it. The German publisher doesn't have on its site.

**Comment by tadeaustria** [comment]
> I learned that there are 2 cards that count _spaces you have filled in your city_ – _Ever Wall_ and _Mayor_. Whatever they are (Critters? Constructions? Events?) and from whichever expansions they are… There's a guy that knows more – @Dajakk.

Yes, these are from the Newleaf expansion. For them, it may could have some impact.
Anyway, I have to fix, adding legendaries to full towns, anyway. The more I think about it, one should not be restricted in the order anyone inserts the cards into his town in the app.

**Comment by Dajakk** [comment]
> The more I think about it, one should not be restricted in the order anyone inserts the cards into his town in the app.

Yeah, I have some bugs to be posted, one regarding legendary wife, one regarding photographer, maybe some more. I will try to post them in the next days

**Comment by tadeaustria** [comment]
> Yeah, I have some bugs to be posted, one regarding legendary wife, one regarding photographer, maybe some more. I will try to post them in the next days
Sure, feel free to post issues that occur. If possible try to separate them, so they can be disussed and solved separately.

**Comment by tadeaustria** [comment]
I've pushed some experimental, quick and dirty implementation of all the mini expansions to https://barthler.ddns.net/EverdellScore.
I am yet not sure, how to group them - separately, grouped as Glimmergold or something different.
I am also not quite happy about the Bridge in the Sky. It has a really unique behaviour and I am not sure how its occupancy works, if the building on top has for example no occupancy.
Selecting Legendary into full town has also not yet been tackled

**Comment by Dajakk** [comment]
> I am yet not sure, how to group them - separately, grouped as Glimmergold or something different.

Actually that was something I wanted to discuss with you. I don't have a final idea crystal clear yet, but...
I was thinking about some modularization of the expansions. There are 3 expansion I think, that work more in a modular way, which means part of it can be included - Bellfaire, Mistwood and Newleaf. I was thinking about changing managing of expansions to more like managing collection, which can be saved in a local storage, and then instead of selecting of expansions, user would select modules he is using. So for example, legends would be a separate module and it would include both Legends 1 and More Legends from Mistwood. Because if you have a complete collection, you don't really care about expansions anymore, you probably have mixed legends from both boxes. This way we could adress all different ways user got some cards. For example in Poland we don't have Glimmergold, we had 3 mini expansion boxes, right now they are out of stock, probably not gonna be reprinted, but they are included in a complete collection.

Similar issue I found is with events, both basic and special. For example I can see myself including just extra events from Newleaf and Bellfaire even without the cards etc. Right now it's necessary to enable all or nothing.

Another thing is with Farms and Mistwood. If you enable Mistwood (for example to use player powers) you are forced to use special farms (only 1 is added since it makes a difference in our case, but probably others should be added too), but normal farms count do not change, so you can basically add 9 farms right now.

**Comment by Dajakk** [comment]
> I've pushed some experimental, quick and dirty implementation of all the mini expansions to https://barthler.ddns.net/EverdellScore.

I will take a look at that branch. For now I've added one comment to it. Maybe let's create a Pool Request to main and have a proper code review?

Additionally... I could add translations for the rest of the cards, but I am not sure if I have rights to push to that branch. When I tried in the past I was not able to push to your repo

**Comment by tadeaustria** [comment]
> Actually that was something I wanted to discuss with you. I don't have a final idea crystal clear yet, but... I was thinking about some modularization of the expansions. There are 3 expansion I think, that work more in a modular way, which means part of it can be included - Bellfaire, Mistwood and Newleaf. I was thinking about changing managing of expansions to more like managing collection, which can be saved in a local storage, and then instead of selecting of expansions, user would select modules he is using. So for example, legends would be a separate module and it would include both Legends 1 and More Legends from Mistwood. Because if you have a complete collection, you don't really care about expansions anymore, you probably have mixed legends from both boxes. This way we could adress all different ways user got some cards. For example in Poland we don't have Glimmergold, we had 3 mini expansion boxes, right now they are out of stock, probably not gonna be reprinted, but they are included in a complete collection.

Yes, very good point. I never though about players with the full collection. But I also was thinking about the more modular expansions, if everything shall be separated. I'm afraid if it gets separated, it could be very complicated to configure a game in the first place.
Maybe we could provide both ways of configuration. Either per Expansion, or per Module.

**Comment by tadeaustria** [comment]
> Similar issue I found is with events, both basic and special. For example I can see myself including just extra events from Newleaf and Bellfaire even without the cards etc. Right now it's necessary to enable all or nothing.

But the grouping of the modules is also quite challenging. I am not sure, how to group all the events. Shall there be a special event group per expansion? One big problem is, that if one configures the game wrongly and wants to enter a event, that isn't there, all data gets lost.

**Comment by tadeaustria** [comment]
> Another thing is with Farms and Mistwood. If you enable Mistwood (for example to use player powers) you are forced to use special farms (only 1 is added since it makes a difference in our case, but probably others should be added too), but normal farms count do not change, so you can basically add 9 farms right now.

Yes, you are right. Right now, with the farms is a small error I intentionally accepted to minimize the programming effort and keep also the UI simple. Would you add all of the different farm seperately?

**Comment by tadeaustria** [comment]
> Additionally... I could add translations for the rest of the cards, but I am not sure if I have rights to push to that branch. When I tried in the past I was not able to push to your repo

I've added you as contributor. So you should now be able to push to non-main branches. Please have a try

**Comment by Dajakk** [comment]
> One big problem is, that if one configures the game wrongly and wants to enter a event, that isn't there, all data gets lost.

Actually enabling most of the things should be safe. Disabling would be problematic for sure, but as long as we are talking about enabling, the only problem I can see is Pearlbrook, and only when any basic event was already added. There could be a warning that either basic events will be removed or that because the settings are not compatible the reset is needed.

> But the grouping of the modules is also quite challenging. I am not sure, how to group all the events. Shall there be a special event group per expansion?

That's why I said, I do not have everything working yet, just a general idea. I don't know, maybe something like this:
- When user enters the app for the first time - display the settings, where he can configure his collection. There he would see all the combinations of boxes he could purchase. So there would be a tickbox for a complete collection, separate checkboxes for all big expansions, for small expansions, for glimmerlock, maybe separate section for promos. Underneath that all would be translated to the smallest chunks, so complete collection would be saved as a collection of everthing except promos, glimmerlock would be stored as extra, legends and rugwort, etc
- Same page could be accessible from the app, settings would be saved in local storage.
- There could be an option to define, what the user considers to be a "base game", for example like @Kaligula0 mentioned, he always plays with Legends and Extra Extra, so he could check it once, and those would be always enabled by default. Same way I would check, that I always play with all the basic/special events from everything but Pearlbrook (as those required special cards)
- Then on the main page we could have section like now for expansions, where you could enable whole boxes, and another section with modules. Obviously enabling for example whole Mistwood would result in enabling modules "legends", "corrin", "farms" (but not player powers probably, those would be enabled by Bellfair)

Of course that's just a draft of how it could work, the last thing I would want is to complicate the user experience too much

**Comment by Kaligula0** [comment]
As for the UI – I imagine this as follows:
 - now there's an expandable section "Expansions" in the left menu; expanded, allows toggling expansions, thus enabling specific features;
 - after the change, enabling an expansion could expand a sub-section of modules which can be toggled (added to the current play), enabling specific features.

When it comes to number of modules in each expansion – it's a matter of arrangements (what can be separated + how hard is it to code + what are the chances of a player using this part only).

**Comment by Kaligula0** [comment]
(UI) … and there may be two checkboxes above the list of expansions: Glimmergold and Complete Collection (anything else? 🤔). Checking one of them would toggle specific expansions that are part of these sets.

**Comment by tadeaustria** [comment]
> Actually enabling most of the things should be safe. Disabling would be problematic for sure, but as long as we are talking about enabling, the only problem I can see is Pearlbrook, and only when any basic event was already added. There could be a warning that either basic events will be removed or that because the settings are not compatible the reset is needed.

Exactely. That's why i chose the current design - Expansions have to be chosen in advance, as espacially disabling would require to cleanup already chosen cards. My universal solution, was to purge everything if anything of the expansions are changed. 
I honestly don't like to distinguish between enabling and disabling, because as you say enabling pearlbrook is also not "safe".
Therefore, I prefer to either to a proper implementation of a "safe" expansion change, where only "removed" content is lost, but I currently don't have the ressources for that. So I would simply keep, that you have to choose your expansions/modules at the start.

**Comment by tadeaustria** [comment]
> Then on the main page we could have section like now for expansions, where you could enable whole boxes, and another section with modules. Obviously enabling for example whole Mistwood would result in enabling modules "legends", "corrin", "farms" (but not player powers probably, those would be enabled by Bellfair)

Most of the player powers are in Bellfair, but the only player power that effects the points i in Mistwood...

**Comment by tadeaustria** [comment]
> That's why I said, I do not have everything working yet, just a general idea. I don't know, maybe something like this:
I could think of an expert tab, additionally to the current expansion tab such that:
* Expansion
  * Bellfair
  * Pearlbrook
  * ...
  * (Complete Collection?)
* Modules (Expert Mode)
  * Bellfair Events
  * Bellfair Basic Event
  * Bellfair Special Event
  * Pearlbrook ???
  * Mistwood Farms
  * Mistwood Legends
  * Mitswood Corrin
  * ...
  * All Legends (Mistwood + Legendary Pack) Maybe too?!?

And my idea would be, if you choose the expansions, all of the related modules are automatically checked, but with the expert tab you could finetune.

**Comment by tadeaustria** [comment]
> Same page could be accessible from the app, settings would be saved in local storage.

Well, I think, what and how to choose the modules or settings is one thing. Saving these settings more permanent is a different feature. But I agree, choosing the modules can become pretty tedious and should be stored for future games.
I am not sure about the idea for a personal _BasicGame_ setting. If you keep the settings from the last game, it should not be so hard to adjust these for the next game IMO.

**Comment by Dajakk** [comment]
> I've added you as contributor. So you should now be able to push to non-main branches. Please have a try

That works great, thanks. I've already added Polish translations for missing cards and created some issues I've found and some we talked about here. When I have some free time I will read that whole discussion and create missing issues to not loose some of the ideas
